src/main/java/org/testng/internal/Invoker.java

```diff
@@ -1123,153 +907,142 @@ private ParameterBag createParameters(ITestNGMethod testMethod,
     final ITestClass testClass= testMethod.getTestClass();
     final List<ITestResult> result = Lists.newArrayList();
     final FailureContext failure = new FailureContext();
-    final ITestNGMethod[] beforeMethods = filterMethods(testClass, testClass.getBeforeTestMethods(), CAN_RUN_FROM_CLASS);
-    final ITestNGMethod[] afterMethods = filterMethods(testClass, testClass.getAfterTestMethods(), CAN_RUN_FROM_CLASS);
-    while(invocationCount-- > 0) {
-      if(false) {
-        // Prevent code formatting
-      }
-      //
-      // No threads, regular invocation
-      //
-      else {
-        // Used in catch statement
-        long start = System.currentTimeMillis();
-
-        Map<String, String> allParameterNames = Maps.newHashMap();
-        ParameterBag bag = createParameters(testMethod,
-            parameters, allParameterNames, suite, testContext, instance);
-
-        if (bag.hasErrors()) {
-          final ITestResult tr = bag.errorResult;
-          Throwable throwable = tr.getThrowable();
-          if (throwable instanceof TestNGException) {
-            tr.setStatus(ITestResult.FAILURE);
-            m_notifier.addFailedTest(testMethod, tr);
-          } else {
-            tr.setStatus(ITestResult.SKIP);
-            m_notifier.addSkippedTest(testMethod, tr);
-          }
-          runTestListeners(tr);
-          result.add(tr);
-          continue;
+    final ITestNGMethod[] beforeMethods = TestNgMethodUtils.filterBeforeTestMethods(testClass, CAN_RUN_FROM_CLASS);
+    final ITestNGMethod[] afterMethods = TestNgMethodUtils.filterAfterTestMethods(testClass, CAN_RUN_FROM_CLASS);
+    while (invocationCount-- > 0) {
```

Comment:
Wouldn't this alter the behavior ? Because the current while loop has already reduced `invocationCount` by 1, before the loop executes, but the for loop would cause the loop to be executed with the current value of `invocationCount` before it gets reduced by 1.

For e.g., Lets say `invocationCount` is 10, the while loop causes the iteration to start with a value of `9`.

The `for` loop would cause it start from `10`.

That would mean, if I put in this change in this PR, it may need regression. I dont want to add too many changes in one PR.

Reply:
true, but we don't use `invocationCount` into the loop, so it should not have a side effect.

Reply:
oops, we are... forget the idea or put `invocationCount--` at the end of the loop.